File: backend/app/models/offer.py
import uuid
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, Enum, ForeignKey
from sqlalchemy.orm import relationship
from ..database import Base
import enum


class OfferStatus(str, enum.Enum):
    pending = "pending"
    accepted = "accepted"
    declined = "declined"
    expired = "expired"


class Offer(Base):
    __tablename__ = "offers"

    id = Column(String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    from_user_id = Column(String(36), ForeignKey("users.id"), nullable=False)
    to_user_id = Column(String(36), ForeignKey("users.id"), nullable=False)
    target_piece_id = Column(String(36), ForeignKey("pieces.id"), nullable=False)
    offered_piece_id = Column(String(36), ForeignKey("pieces.id"), nullable=False)
    offered_pinceaux = Column(Integer, default=0)
    status = Column(Enum(OfferStatus), default=OfferStatus.pending)
    created_at = Column(DateTime, default=datetime.utcnow)

    from_user = relationship("User", foreign_keys=[from_user_id], back_populates="offers_made")
    to_user = relationship("User", foreign_keys=[to_user_id], back_populates="offers_received")

File: BACKUP002/mobile/lib/models/offer.dart
class Offer {
  final String id;
  final String fromUserId;
  final String toUserId;
  final String targetPieceId;
  final String offeredPieceId;
  final int offeredPinceaux;
  final String status;
  final DateTime createdAt;

  Offer({
    required this.id,
    required this.fromUserId,
    required this.toUserId,
    required this.targetPieceId,
    required this.offeredPieceId,
    this.offeredPinceaux = 0,
    this.status = 'pending',
    DateTime? createdAt,
  }) : createdAt = createdAt ?? DateTime.now();

  factory Offer.fromJson(Map<String, dynamic> json) {
    return Offer(
      id: json['id'] as String,
      fromUserId: json['from_user_id'] as String,
      toUserId: json['to_user_id'] as String,
      targetPieceId: json['target_piece_id'] as String,
      offeredPieceId: json['offered_piece_id'] as String,
      offeredPinceaux: (json['offered_pinceaux'] as int?) ?? 0,
      status: (json['status'] as String?) ?? 'pending',
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }
}
